libs/@guardian/identity-auth/src/token.ts

```diff
@@ -0,0 +1,910 @@
+
+		// add the response handler to the window
+		window.addEventListener('message', responseHandler);
+
+		// set the timeout
+		timeoutId = window.setTimeout(() => {
+			// reject the promise with a timeout error if the timeout is reached
+			return reject(
+				new OAuthError({
+					error: 'timeout',
+					error_description: 'Timeout',
+					message: 'The oauth request timed out',
+				}),
+			);
+		}, timeout);
+	});
+
+	// return the promise, clearing the timeout and removing the event listener when the promise resolves
+	return msgReceivedOrTimeout.finally(() => {
+		clearTimeout(timeoutId);
+		window.removeEventListener('message', responseHandler);
+	});
+};
+
+/**
+ * @name loadFrame
+ * @description Loads a hidden iframe with the given url
+ *
+ * @param url - url to load in the iframe
+ * @returns	HTMLIFrameElement - the iframe element
+ */
+export const loadFrame = (url: string) => {
+	const frame = document.createElement('iframe');
+	frame.style.display = 'none';
+	frame.src = url;
+
+	return document.body.appendChild(frame);
+};
+
+/**
+ * @name performAuthCodeFlowIframe
+ * @description Performs the authorization code flow with PKCE using an iframe, and returns the AuthorizationResponse
+ *
+ * @param authorizeParams - AuthorizeParams
+ * @param options - IdentityAuthOptions
+ * @param oauthUrls - OAuthUrls
+ * @returns OAuthAuthorizeResponse | OAuthAuthorizeResponseError - resolves with the AuthorizationResponse
+ */
+export const performAuthCodeFlowIframe = async (
+	authorizeParams: AuthorizeParams,
+	options: IdentityAuthOptions,
+	oauthUrls: OAuthUrls,
+): Promise<OAuthAuthorizeResponse | OAuthAuthorizeResponseError> => {
+	// convert the authorize params to a query string
+	const searchParams = new URLSearchParams(authorizeParams);
+
+	// create the postMessage listener for the iframe
+	const postMessageListener = addPostMessageListener(
+		options,
+		authorizeParams.state,
+	);
+	// load the iframe with the authorize url
+	const iframe = loadFrame(
+		`${oauthUrls.authorizeUrl}?${searchParams.toString()}`,
+	);
+
+	// wait for the postMessage listener to resolve
+	const authorizeResponse = await postMessageListener.finally(() => {
+		if (document.body.contains(iframe)) {
+			document.body.removeChild(iframe);
+		}
+	});
+
+	return authorizeResponse;
+};
+
+// memoized verify token map
+const memoizedVerifyToken = new Map<string, unknown>();
+
+/**
+ * @name verifyToken
+ * @description Verifies the ID token, checking the signature and claims, and verifies the access token using the at_hash claim
+ *
+ * @param idToken - IDToken
+ * @param accessToken - AccessToken
+ * @param options - IdentityAuthOptions
+ * @returns Promise<void> - resolves when the token is verified
+ */
+export const verifyToken = async (
+	idToken: IDToken,
+	accessToken: AccessToken,
+	options: IdentityAuthOptions,
+	oauthUrls: OAuthUrls,
+) => {
+	// check if the token has already been verified
+	if (memoizedVerifyToken.has(idToken.claims.jti)) {
```

Comment:
Is there any potential here for getting around verification by marking a token as verified when it hasn't been?  I know that's a slightly vague question as I'm not sure how the memoization is working

Reply:
i dont think so, it's stored in a `Map`, where the only reference to it is in the method below that, so there isn't any other way of getting to that Map in memory.